src/client/actions.ts

```diff
@@ -0,0 +1,132 @@
+ *             iconUrl: 'https://www.example.com/timer.png',
+ *             onClick: {
+ *                 action: 'schedule-reminder',
+ *                 intervalMs: 5 * 60 * 1000
+ *             }
+ *         }
+ *     ]
+ * });
+ *
+ * // Create a listener that will be called for each action
+ * // Note: This handler will be used for ALL actions, from ALL notifications that are created by this application.
+ * addEventListener('notification-action', (event: NotificationActionEvent) => {
+ *     const {result, notification} = event;
+ *
+ *     if (result.action === 'view-event') {
+ *         // Open a window with full details of the associated event
+ *         openEventDetails(notification.customData.eventId, result.target);
+ *     } else if (result.action === 'schedule-reminder') {
+ *         // Schedule a new notification
+ *         scheduleReminder(notification.customData.eventId, Date.now() + result.intervalMs);
+ *     } // Etc...
+ * });
+ * ```
+ *
+ * The example above uses `customData` to store details about the notification subject (in this case, a calendar
+ * event), and `onClick` actions to inform the application about how it should respond when the user interacts with the
+ * notification. This is our intended usage and recommended best-practice, but the service doesn't require applications
+ * to follow this convention - application developers are free to decide how to manage notification state.
+ *
+ * Within the `notification-action` handler, the application must be able to understand which notification is being
+ * handled, and how to decide what it should do next. The example above uses an application-defined `action` field to
+ * determine the correct action, but the notification's `id`, `category`, `customData` and other fields are also useful
+ * selectors.
+ *
+ * @module Actions
+ */
+
+/**
+ * Denotes a field as being an action. Defining this field (with a non-`undefined` value) will result in actions being
+ * generated and sent back to the source application when the corresponding event happens.
+ *
+ * For example, providing a value for the `onClick` field of {@link ButtonOptions} will result in a
+ * `notification-action` event being raised when that button is clicked.
+ *
+ * In the current version of the service, the result payloads returned back to an application are static and must be
+ * defined at the point where the notification is created. Later versions of the service will allow some limited
+ * programmatic creation of these payloads, for use in situations where a static payload isn't sufficient.
+ *
+ * *The generic parameters of this type are for future expansion. Future versions of the service will allow for more
+ * control over the handling of actions.*
+ */
+export type ActionDeclaration<T extends never, E extends never> = NotificationActionResult;
+
+/**
+ * Data type used to represent the payloads returned back to applications as the result of an action. Applications
```

Comment:
'Data type used to represent the action result returned back to applications when an action is raised'?